### PR TITLE
#28 EA-1988, EA-1987 implemented open/closed sets handling to support

### PR DESCRIPTION
**Pull request by rgraf**
curated datasets

### CODE REVIEW

File: set-web/src/main/java/eu/europeana/set/web/service/UserSetService.java [hunk 1]
```diff
@@ -21,8 +21,28 @@
...
+    		String sort, String sortOrder, int pageNr, int pageSize)
     	    throws HttpException, IOException, JSONException;
+    
+    /**
+     * This method converts open set to close set by updating respective items
+     * @param storedUserSet
+     * @param items
+     * @return updated set
+     */
+    public UserSet updateUserSetsWithCloseSetItems(UserSet storedUserSet, List<String> items);
```
[the remark this answers is not in the document]
Comment (gsergiu) [reply set-web/src/main/java/eu/europeana/set/web/service/UserSetService.java:44]: the difference is in the parameters

File: set-web/src/main/java/eu/europeana/set/web/service/controller/BaseRest.java [hunk 2]
```diff
@@ -188,6 +184,29 @@ public LdProfiles getProfile(String paramProfile, HttpServletRequest request) th
...
+	 * @throws HttpException
+	 * @throws UserSetProfileValidationException
+	 */
+	public LdProfiles getHeaderProfile(HttpServletRequest request) throws HttpException {
+
+		LdProfiles profile = null;
+		String preferHeader = request.getHeader(HttpHeaders.PREFER);
+		if (preferHeader != null) {
+			// identify profile by prefer header
+			profile = getProfile(preferHeader);
```
Comment (gsergiu) [inline set-web/src/main/java/eu/europeana/set/web/service/controller/BaseRest.java:204]: we don't need two different methos getHeaderProfile and getProfile

File: set-web/src/main/java/eu/europeana/set/web/service/controller/BaseRest.java [hunk 3]
```diff
@@ -391,13 +410,16 @@ public String getApiVersion() {
      * @throws IOException
      * @throws JSONException
      */
-    public UserSet updateItemsWithIsDefinedBy(UserSet userSet) throws HttpException, IOException, JSONException {
-		if (userSet.getIsDefinedBy() != null) {
+    public UserSet updateItemsWithIsDefinedBy(UserSet userSet,
```
Comment (gsergiu) [inline set-web/src/main/java/eu/europeana/set/web/service/controller/BaseRest.java:413]: please rename to fetchItemsPage

File: set-web/src/main/java/eu/europeana/set/web/service/controller/exception/GlobalExceptionHandler.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+	@Resource
+	I18nService i18nService;
+
+	protected I18nService getI18nService() {
+		return i18nService;
+	}
+	
+	@Override
+	public ApiResponse buildErrorResponse(String errorMessage, String action, String apiKey) {
```
Comment (gsergiu) [inline set-web/src/main/java/eu/europeana/set/web/service/controller/exception/GlobalExceptionHandler.java:23]: Why is this method needed? There is an buildErrorResponsemethod in the paren implementation

File: set-web/src/main/java/eu/europeana/set/web/service/controller/jsonld/WebUserSetRest.java [hunk 5]
```diff
@@ -279,6 +287,32 @@
...
+							new String[] {}, HttpStatus.PRECONDITION_FAILED, null);	
+					}
+
+					// if the Set corresponds to a closed set, replace member items with the new items 
+					// that are present in the Set description only when a profile is indicated and 
+					// is different from "ldp:PreferMinimalContainer" is referred in the "Prefer" header.
+					if (getHeaderProfile(request) != LdProfiles.MINIMAL)
+						newUserSet = getUserSetService().updateUserSetsWithCloseSetItems(existingUserSet, newUserSet.getItems());
+				}
+
```
Comment (gsergiu) [inline set-web/src/main/java/eu/europeana/set/web/service/controller/jsonld/WebUserSetRest.java:315]: the generation of the eTag should happen after updating the user set into database, see also the updates to specification

File: set-web/src/main/java/eu/europeana/set/web/service/impl/UserSetServiceImpl.java [hunk 6]
```diff
@@ -260,6 +267,13 @@ public void validateWebUserSet(UserSet webUserSet) throws RequestBodyValidationE
 	 */
 	public void deleteUserSet(String userSetId) throws UserSetNotFoundException {
 
+		// in case it is a closed set, remove the items that are members of the Set.
+		UserSet userSet = getUserSetById(userSetId);
+		if (isOpenSet(userSet)) {
```
Comment (gsergiu) [inline set-web/src/main/java/eu/europeana/set/web/service/impl/UserSetServiceImpl.java:272]: this code is note doing anything, please remove it and double check if the reading of the UserSet from database is needed at this stage

File: set-web/src/main/java/eu/europeana/set/web/service/impl/UserSetServiceImpl.java [hunk 7]
```diff
@@ -374,17 +388,85 @@ public void addNewItemToList(UserSet existingUserSet, int positionInt, String ne
 	}
 
     @Override
-    public UserSet updateUserSetsWithIsDefinedByUrl(UserSet storedUserSet, String apiKey, String action)
+    public UserSet updateUserSetsWithIsDefinedByUrl(UserSet storedUserSet, String apiKey, String action,
```
Comment (gsergiu) [inline set-web/src/main/java/eu/europeana/set/web/service/impl/UserSetServiceImpl.java:391]: please rename to updateUserSet or updateOpenSet if required

Comment (gsergiu) [reply set-web/src/main/java/eu/europeana/set/web/service/impl/UserSetServiceImpl.java:391]: isn't this the same functionality as the fetchItems in the controller? Please check if code dupplication was introduced!

File: set-web/src/main/java/eu/europeana/set/web/service/impl/UserSetServiceImpl.java [hunk 8]
```diff
@@ -374,17 +388,85 @@ public void addNewItemToList(UserSet existingUserSet, int positionInt, String ne
...
+        
+    /* (non-Javadoc)
+     * @see eu.europeana.set.web.service.UserSetService#updateUserSetsWithCloseSetItems(eu.europeana.set.definitions.model.UserSet, java.util.List)
+     */
+    public UserSet updateUserSetsWithCloseSetItems(UserSet storedUserSet, List<String> items) {    	
+    	if (items.size() > 0) {
+	    	storedUserSet.setItems(items);
+	    	storedUserSet.setTotal(items.size());
+    	}
+    	storedUserSet.setIsDefinedBy(null);
```
Comment (gsergiu) [inline set-web/src/main/java/eu/europeana/set/web/service/impl/UserSetServiceImpl.java:458]: why is this needed? The validation rules should prevent having an isDefinedBy value at this stage

File: set-web/src/main/java/eu/europeana/set/web/service/impl/UserSetServiceImpl.java [hunk 9]
```diff
@@ -374,17 +388,85 @@ public void addNewItemToList(UserSet existingUserSet, int positionInt, String ne
...
+			searchQuery.append(sortOrder);
+		}
+
+		return searchQuery.toString();
+	}
+        
+    /* (non-Javadoc)
+     * @see eu.europeana.set.web.service.UserSetService#updateUserSetsWithCloseSetItems(eu.europeana.set.definitions.model.UserSet, java.util.List)
+     */
+    public UserSet updateUserSetsWithCloseSetItems(UserSet storedUserSet, List<String> items) {
```
Comment (gsergiu) [inline set-web/src/main/java/eu/europeana/set/web/service/impl/UserSetServiceImpl.java:453]: is this method really needed? It only sets 2 values in the userSet

File: set-web/src/main/java/eu/europeana/set/web/service/impl/UserSetServiceImpl.java [hunk 10]
```diff
@@ -374,17 +388,85 @@ public void addNewItemToList(UserSet existingUserSet, int positionInt, String ne
...
+	    	storedUserSet.setTotal(items.size());
+    	}
+    	storedUserSet.setIsDefinedBy(null);
+    	return storedUserSet;
+    }
+        
+    /* (non-Javadoc)
+     * @see eu.europeana.set.web.service.UserSetService#isOpenSet(eu.europeana.set.definitions.model.UserSet)
+     */
+    public boolean isOpenSet(UserSet userSet) {
```
Comment (gsergiu) [inline set-web/src/main/java/eu/europeana/set/web/service/impl/UserSetServiceImpl.java:465]: please move this methos to UserSet interface and implementation

File: set-web/src/main/java/eu/europeana/set/web/service/impl/UserSetServiceImpl.java (hunk 9 above)
Comment (rgraf) [reply set-web/src/main/java/eu/europeana/set/web/service/impl/UserSetServiceImpl.java:453]: I believe, it is required to support application logic for open/closed sets according to documentation

File: set-web/src/main/java/eu/europeana/set/web/service/impl/UserSetServiceImpl.java (hunk 7 above)
Comment (rgraf) [reply set-web/src/main/java/eu/europeana/set/web/service/impl/UserSetServiceImpl.java:391]: I believe, there is no code duplication. The second parameter of updateUserSet is apikey. The method fetchItems prepares this parameter and calls updateUserSet method.

File: set-web/src/main/java/eu/europeana/set/web/service/UserSetService.java [hunk 11]
```diff
@@ -21,8 +21,28 @@
...
+     * This method converts close set to open set by updating respective items
+     * @param storedUserSet
+     * @param apiKey
+     * @param action
+     * @return updated set
+     * @throws HttpException
+     * @throws IOException
+     * @throws JSONException
+     */
+    public UserSet updateUserSetsWithIsDefinedByUrl(UserSet storedUserSet, String apiKey, String action,
```
[the remark this answers is not in the document]
Comment (rgraf) [reply set-web/src/main/java/eu/europeana/set/web/service/UserSetService.java]: method renamed as required

File: set-web/src/main/java/eu/europeana/set/web/service/UserSetService.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (rgraf) [reply set-web/src/main/java/eu/europeana/set/web/service/UserSetService.java]: method renamed as required

File: set-web/src/main/java/eu/europeana/set/web/service/UserSetService.java [hunk 12]
```diff
@@ -159,4 +179,12 @@ public UserSet insertItem(String datasetId, String localId, String position, Use
 	 * @param newItem
 	 */
 	public void addNewItemToList(UserSet existingUserSet, int positionInt, String newItem);	
+	
+    /**
+     * This method detects if it is open or closed set
+     * @param userSet
+     * @return true if it is an open set
+     */
+    public boolean isOpenSet(UserSet userSet);
```
[the remark this answers is not in the document]
Comment (rgraf) [reply set-web/src/main/java/eu/europeana/set/web/service/UserSetService.java]: method moved as required

File: set-web/src/main/java/eu/europeana/set/web/service/controller/BaseRest.java (hunk 2 above)
Comment (rgraf) [reply set-web/src/main/java/eu/europeana/set/web/service/controller/BaseRest.java]: combined in one method

File: set-web/src/main/java/eu/europeana/set/web/service/controller/BaseRest.java (hunk 3 above)
Comment (rgraf) [reply set-web/src/main/java/eu/europeana/set/web/service/controller/BaseRest.java]: renamed as required

File: set-web/src/main/java/eu/europeana/set/web/service/controller/exception/GlobalExceptionHandler.java (hunk 4 above)
Comment (rgraf) [reply set-web/src/main/java/eu/europeana/set/web/service/controller/exception/GlobalExceptionHandler.java]: removed as suggested

File: set-web/src/main/java/eu/europeana/set/web/service/controller/jsonld/WebUserSetRest.java (hunk 5 above)
Comment (rgraf) [reply set-web/src/main/java/eu/europeana/set/web/service/controller/jsonld/WebUserSetRest.java:315]: improved logic according to suggestion and specification

File: set-web/src/main/java/eu/europeana/set/web/service/impl/UserSetServiceImpl.java (hunk 6 above)
Comment (rgraf) [reply set-web/src/main/java/eu/europeana/set/web/service/impl/UserSetServiceImpl.java]: method refactored according to specifications

File: set-web/src/main/java/eu/europeana/set/web/service/impl/UserSetServiceImpl.java (hunk 8 above)
Comment (rgraf) [reply set-web/src/main/java/eu/europeana/set/web/service/impl/UserSetServiceImpl.java]: removed obsolete part

File: set-web/src/main/java/eu/europeana/set/web/service/impl/UserSetServiceImpl.java (hunk 10 above)
Comment (rgraf) [reply set-web/src/main/java/eu/europeana/set/web/service/impl/UserSetServiceImpl.java]: moved as required
